runtime/src/iree/base/tree.c

```diff
@@ -0,0 +1,549 @@
+        t->parent->parent->red = true;
+        t = t->parent->parent;
+      } else {
+        if (t == t->parent->right) {
+          t = t->parent;
+          iree_tree_rotate_left(tree, t);
+        }
+        t->parent->red = false;
+        t->parent->parent->red = true;
+        iree_tree_rotate_right(tree, t->parent->parent);
+      }
+    }
+    if (t == tree->root) {
+      break;
+    }
+  }
+  tree->root->red = false;
+
+  return iree_ok_status();
+}
+
+iree_status_t iree_tree_insert(iree_tree_t* tree, iree_host_size_t key,
+                               iree_tree_node_t** out_data) {
+  iree_tree_node_t* t;
+  iree_status_t status = iree_tree_allocate_node(tree, &t);
+  if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+    return status;
+  }
+  status = iree_tree_insert_internal(tree, key, t);
+  if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
+    return status;
+  }
+  tree->size++;
+  *out_data = t;
+  return status;
+}
+
+iree_tree_node_t* iree_tree_get(iree_tree_t* tree, iree_host_size_t key) {
+  iree_tree_node_t* node = tree->root;
+  while (node->is_sentinel == false) {
+    if (key == node->key) {
+      return node;
+    } else if (key < node->key) {
+      node = node->left;
+    } else {
+      node = node->right;
+    }
+  }
+  return NULL;
+}
+
+static bool iree_tree_walk_helper(iree_tree_node_t* node,
+                                  iree_tree_walk_type_e walk_type,
+                                  bool (*callback)(iree_tree_node_t*, void*),
+                                  void* user_data) {
+  IREE_ASSERT(walk_type <= IREE_TREE_WALK_POSTORDER);
+  if (!node || node->is_sentinel) {
+    return true;
+  }
+  switch (walk_type) {
+    case IREE_TREE_WALK_PREORDER:
+      if (!callback(node, user_data)) {
+        return false;
+      }
+      if (!iree_tree_walk_helper(node->left, walk_type, callback, user_data)) {
+        return false;
+      }
+      return iree_tree_walk_helper(node->right, walk_type, callback, user_data);
+    case IREE_TREE_WALK_INORDER:
+      if (!iree_tree_walk_helper(node->left, walk_type, callback, user_data)) {
+        return false;
+      }
+      if (!callback(node, user_data)) {
+        return false;
+      }
+      return iree_tree_walk_helper(node->right, walk_type, callback, user_data);
+    case IREE_TREE_WALK_POSTORDER:
+      if (!iree_tree_walk_helper(node->left, walk_type, callback, user_data)) {
+        return false;
+      }
+      if (!iree_tree_walk_helper(node->right, walk_type, callback, user_data)) {
+        return false;
+      }
+      return callback(node, user_data);
+  }
+  return false;
+}
+
+void iree_tree_walk(iree_tree_t* tree, iree_tree_walk_type_e walk_type,
+                    bool (*callback)(iree_tree_node_t*, void*),
+                    void* user_data) {
+  iree_tree_walk_helper(tree->root, walk_type, callback, user_data);
+}
+
+iree_tree_node_t* iree_tree_node_next(iree_tree_node_t* node) {
+  IREE_ASSERT(node != NULL);
```

Comment:
IREE_ASSERT_ARGUMENT (here and elsewhere in the PR - there's a few styles that are used that should be switched to it)

IREE_ASSERT_ARGUMENT is technically just an assert but that's due to laziness - the intent is to have it emit a better assertion message ("iree_tree_node_next argument `node` is NULL") - it's a 5 year TODO to go do the `__FUNCTION__` stuff - but when we do it'll be nice :)